### PR TITLE
Bug 1352568 - Drop LeanPlum events to serve as triggers for Focus Promotion

### PR DESCRIPTION
**Pull request by bkmunar**

### CODE REVIEW

**Comment by farhanpatel** [comment]
### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
LeanplumIntegration.swift | 88 | Explicitly calling .init() should be avoided. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Client/Application/AppDelegate.swift [hunk 1]
```diff
@@ -157,12 +157,20 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
         
         adjustIntegration = AdjustIntegration(profile: profile)
 
+        LeanplumIntegration.sharedInstance.setup(profile: profile)
+
         // We need to check if the app is a clean install to use for
         // preventing the What's New URL from appearing.
         if getProfile(application).prefs.intForKey(IntroViewControllerSeenProfileKey) == nil {
             getProfile(application).prefs.setString(AppInfo.appVersion, forKey: LatestAppVersionProfileKey)
+            LeanplumIntegration.sharedInstance.track(event: "Opened App - First Run")
```
Comment (Sdaswani) [inline Client/Application/AppDelegate.swift]: Let's just call it 'First Run'.

File: Client/Application/AppDelegate.swift [hunk 2]
```diff
@@ -157,12 +157,20 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
...
+        } else if getProfile(application).prefs.boolForKey("SecondRun") == nil {
+            getProfile(application).prefs.setBool(true, forKey: "SecondRun")
+            LeanplumIntegration.sharedInstance.track(event: "Opened App - Second Run")
```
Comment (Sdaswani) [inline Client/Application/AppDelegate.swift]: Let's just call it 'Second Run'.

File: Client/Entitlements/Debug.entitlements [hunk 3]
```diff
@@ -4,6 +4,7 @@
 <dict>
 	<key>aps-environment</key>
 	<string>development</string>
+
```
Comment (Sdaswani) [inline Client/Entitlements/Debug.entitlements]: Do we need this change?

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 4]
```diff
@@ -1336,6 +1336,8 @@ extension BrowserViewController: MenuActionDelegate {
             switch menuAction {
             case .openNewNormalTab:
                 self.openURLInNewTab(nil, isPrivate: false, isPrivileged: true)
+                LeanplumIntegration.sharedInstance.track(event: "Open new tab")
```
Comment (Sdaswani) [inline Client/Frontend/Browser/BrowserViewController.swift]: Let's be consistent with naming the events - Normal english, but all words capitalized. So 'Open New Tab' instead.

File: Client/Frontend/Settings/AppSettingsOptions.swift [hunk 5]
```diff
@@ -602,8 +602,10 @@ class LoginsSetting: Setting {
     }
 
     override func onClick(_: UINavigationController?) {
+
```
Comment (Sdaswani) [inline Client/Frontend/Settings/AppSettingsOptions.swift]: Remove this whitespace?

**Comment by farhanpatel** [comment]
@bkmunar have you looked at adding leanplum via carthage? 
https://github.com/Leanplum/Leanplum-iOS-SDK

It'll keep the changes to the project smaller :)

**Comment by bkmunar** [comment]
can't use it -- it's closed source. we're blocked on them right now so this is a temporary fix that stefan did a while back so i can at least do some work on leanplum.

**Comment by Sdaswani** [comment]
@farhanpatel we have to use an open source SDK, a beta version of which you can find here: https://bugzilla.mozilla.org/show_bug.cgi?id=1353795

Any advice on importing it via Carthage or Cocoapods? I haven't had a chance to dig in and see how we can swap out the closed source version and put this one in.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 6]
```diff
@@ -2295,6 +2295,8 @@ extension BrowserViewController: WKNavigationDelegate {
                     UIApplication.shared.openURL(url)
                 }))
                 present(alert, animated: true, completion: nil)
+
+                LeanplumIntegration.sharedInstance.track(event: "Open telephone link")
```
Comment (Sdaswani) [inline Client/Frontend/Browser/BrowserViewController.swift]: Let's follow all upper case with LP events - so 'Open Telephone Link'.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 7]
```diff
@@ -2317,6 +2319,8 @@ extension BrowserViewController: WKNavigationDelegate {
             } else {
                 UIApplication.shared.openURL(url)
             }
+
+            LeanplumIntegration.sharedInstance.track(event: "Open mailto link")
```
Comment (Sdaswani) [inline Client/Frontend/Browser/BrowserViewController.swift]: Let's make it 'Open Mailto Link'

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 8]
```diff
@@ -3032,7 +3036,7 @@ extension BrowserViewController: ContextMenuHelperDelegate {
                     }
                     accessDenied.addAction(settingsAction)
                     self.present(accessDenied, animated: true, completion: nil)
-
+                    LeanplumIntegration.sharedInstance.track(event: "Download media - Saved Image")
```
Comment (Sdaswani) [inline Client/Frontend/Browser/BrowserViewController.swift]: Let's make it 'Download Media - Saved Image'

File: Client/Frontend/Browser/LoginsHelper.swift [hunk 9]
```diff
@@ -172,6 +172,8 @@ class LoginsHelper: TabHelper {
                     self.tab?.removeSnackbar(bar)
                     self.snackBar = nil
                     self.profile.logins.addLogin(login)
+
+                    LeanplumIntegration.sharedInstance.track(event: "Save login and password")
```
Comment (Sdaswani) [inline Client/Frontend/Browser/LoginsHelper.swift]: Let's make it 'Save Login and Password'

File: Client/Frontend/Browser/TabTrayController.swift [hunk 10]
```diff
@@ -569,6 +569,8 @@ class TabTrayController: UIViewController {
         let exitingPrivateMode = !privateMode && profile.prefs.boolForKey("settings.closePrivateTabs") ?? false
         if exitingPrivateMode {
             tabManager.removeAllPrivateTabsAndNotify(false)
+
+            LeanplumIntegration.sharedInstance.track(event: "Close Private Tabs when leaving Private Browsing")
```
Comment (Sdaswani) [inline Client/Frontend/Browser/TabTrayController.swift]: Let's make it "Close Private Tabs When Leaving Private Browsing"

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 11]
```diff
@@ -1624,6 +1622,8 @@ extension BrowserViewController: URLBarDelegate {
             }
             showHomePanelController(inline: false)
         }
+
+        LeanplumIntegration.sharedInstance.track(event: "Interact with search url area")
```
Comment (Sdaswani) [inline Client/Frontend/Browser/BrowserViewController.swift]: Let's make it "Interact With Search URL Area"

File: Client/Frontend/Browser/TabManager.swift [hunk 12]
```diff
@@ -453,14 +453,38 @@ class TabManager: NSObject {
...
+        // add check for event stuff
+        guard let shouldShowFocusUI = LeanplumIntegration.sharedInstance.getBoolVariableFromServer(key: "shouldShowFocusUI"), let focus = NSURL(string: "firefox-focus://"), let klar = NSURL(string: "firefox-klar://"), !UIApplication.shared.canOpenURL(focus as URL) && !UIApplication.shared.canOpenURL(klar as URL) && shouldShowFocusUI else {
+            return
+        }
+
+
+        var toast: LeanplumPromoButtonToast
+        let templateDict = LeanplumIntegration.sharedInstance.getTemplateDictionary()!
+        toast = LeanplumPromoButtonToast(labelText: templateDict["Template Text"]! , buttonText: templateDict["Button Text"]!, colorText: templateDict["Hex Color String"]!, completion: { buttonPressed in
+            guard let url = NSURL(string: "https://itunes.apple.com/app/id1055677337") else {
```
Comment (Sdaswani) [inline Client/Frontend/Browser/TabManager.swift]: Why are you confining the URL here to this URL? It should be set in the template dictionary?

File: Client/Frontend/Strings.swift [hunk 13]
```diff
@@ -280,6 +280,12 @@ extension Strings {
     public static let ReaderModeResetFontSizeAccessibilityLabel = NSLocalizedString("Reset text size", comment: "Accessibility label for button resetting font size in display settings of reader mode")
 }
 
+// Leanplum Strings.
+extension Strings {
```
Comment (Sdaswani) [inline Client/Frontend/Strings.swift]: I don't think we need these String, right? Everything comes from LP so we don't need to localize anything?

File: Client/Frontend/Browser/ButtonToast.swift [hunk 14]
```diff
@@ -113,62 +113,118 @@ class ButtonToast: UIView {
...
+
+class LeanplumPromoButtonToast : ButtonToast {
+    let colorText: String
+
+    init(labelText: String, descriptionText: String? = nil, buttonText: String, colorText: String, completion:@escaping (_ buttonPressed: Bool) -> Void) {
+        self.colorText = colorText
+        super.init(labelText: labelText, descriptionText: descriptionText, buttonText: buttonText, completion: completion)
+
+        self.snp.remakeConstraints { make in
+            make.height.equalTo(96)
```
Comment (farhanpatel) [inline Client/Frontend/Browser/ButtonToast.swift]: static constant instead of just the number  🙃

Comment (farhanpatel) [inline Client/Frontend/Browser/ButtonToast.swift]: same goes for all the numbers in this class.

File: Client/Frontend/Browser/TabManager.swift [hunk 15]
```diff
@@ -420,18 +420,18 @@ class TabManager: NSObject {
         // without this we get a nasty crash
         guard let selectedTab = selectedTab,
             let _ = privateTabs.index(of: selectedTab) else {
-            return privateTabs.forEach({ removeTab($0, flushToDisk: true, notify: notify) })
+                return privateTabs.forEach({ removeTab($0, flushToDisk: true, notify: notify) })
```
Comment (farhanpatel) [inline Client/Frontend/Browser/TabManager.swift]: Same thing goes with all the space changes here.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 16]
```diff
@@ -2353,18 +2357,18 @@ extension BrowserViewController: WKNavigationDelegate {
         // accepted by the user.
         let origin = "\(challenge.protectionSpace.host):\(challenge.protectionSpace.port)"
         if challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
-           let trust = challenge.protectionSpace.serverTrust,
-           let cert = SecTrustGetCertificateAtIndex(trust, 0), profile.certStore.containsCertificate(cert, forOrigin: origin) {
+            let trust = challenge.protectionSpace.serverTrust,
+            let cert = SecTrustGetCertificateAtIndex(trust, 0), profile.certStore.containsCertificate(cert, forOrigin: origin) {
```
Comment (farhanpatel) [inline Client/Frontend/Browser/BrowserViewController.swift]: All the spaces make it hard to see your changes. ☹️

**Comment by farhanpatel** [comment]
Sorry for taking so long to get to this.
2 overall things you should look at. 
- There are a lot of whitespace changes which makes it hard to track whats going on in your PR. Please make sure you remove all of them. I think there might be a setting in your IDE that is doing this? If the whitespace is actually wrong in all these places, its better to make a separate PR for it instead of doing it here. 
- You could use an enum for the events/state strings instead. It'll make sure a typo doesnt mess up your events. Look at how Steph did it here. https://github.com/mozilla-mobile/firefox-ios/blob/master/Client/Frontend/Home/ActivityStreamPanel.swift#L615

**Comment by bkmunar** [comment]
Sorry @farhanpatel i didn't flag you for review did i? I thought stefan was supposed to take a look at it

**Comment by bkmunar** [comment]
Just kidding! @farhanpatel susheel just got back to me about you taking the review instead :)

**Comment by Sdaswani** [comment]
@farhanpatel I agree the whitespace changes make it very difficult to review this easily. @bkmunar I think you said it was part of @st3fan patch for the initial LP inclusion but if you can remove them it would be best.

File: Client/Application/AppDelegate.swift [hunk 17]
```diff
@@ -157,12 +157,20 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
         
         adjustIntegration = AdjustIntegration(profile: profile)
 
+        LeanplumIntegration.sharedInstance.setup(profile: profile)
+
         // We need to check if the app is a clean install to use for
         // preventing the What's New URL from appearing.
         if getProfile(application).prefs.intForKey(IntroViewControllerSeenProfileKey) == nil {
             getProfile(application).prefs.setString(AppInfo.appVersion, forKey: LatestAppVersionProfileKey)
+            LeanplumIntegration.sharedInstance.track(event: "First Run")
```
Comment (st3fan) [inline Client/Application/AppDelegate.swift]: Is it a good idea to have spaces in the event names?

File: Client/Application/LeanplumIntegration.swift [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+        }
+        Leanplum.syncResourcesAsync(true)
+        setupTemplateDictionary()
+
+        var userAttributesDict = [AnyHashable : Any]()
+        userAttributesDict["Focus Installed"] = "false"
+        userAttributesDict["Klar Installed"] = "false"
+        userAttributesDict["Alternate Mail Client Installed"] = "mailto:"
+
+        if let focusURL = NSURL(string: "firefox-focus://"), UIApplication.shared.canOpenURL(focusURL as URL) {
```
Comment (st3fan) [inline Client/Application/LeanplumIntegration.swift]: You can avoid the `as URL` if you create a `URL(string: ...)` I think.

File: Client/Application/LeanplumIntegration.swift [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+        if let mailtoScheme = profile.prefs.stringForKey(PrefsKeys.KeyMailToOption), mailtoScheme != "mailto:" {
+            userAttributesDict["Alternate Mail Client Installed"] = mailtoScheme
+        }
+
+        Leanplum.start(userAttributes: userAttributesDict)
+    }
+
+    // Events
+
+    func track(event: String) {
```
Comment (st3fan) [inline Client/Application/LeanplumIntegration.swift]: How about instead of using a `String` here, use an enum. Then it is more strictly controlled and there will also be  less chance of typos.

We do the same for the Adjust integration in Focus. See:

- https://github.com/mozilla-mobile/focus/blob/master/Blockzilla/AdjustIntegration.swift#L8
- https://github.com/mozilla-mobile/focus/blob/master/Blockzilla/AdjustIntegration.swift#L93

File: Client/Application/AppDelegate.swift (hunk 17 above)
Comment (Sdaswani) [inline Client/Application/AppDelegate.swift]: Stefan, since the events are consumed by Product/Marketing, in the past I've recommended human readable names. The important thing is that we are consistent and we advertise a 'legend' to Product/Marketing (so that is why an enum would help, as @farhanpatel suggested), so if you feel strongly about no spaces, please let us know.

File: Client/Configuration/Fennec.xcconfig [hunk 20]
```diff
@@ -7,6 +7,20 @@ MOZ_BUNDLE_DISPLAY_NAME = Fennec ($(USER))
...
+
+// Flag to indicate if we want to include the debug settings bundle or not
+INCLUDE_SETTINGS_BUNDLE = YES
+
+// Adjust SDK Variables
+ADJUST_ENVIRONMENT =
+ADJUST_APP_TOKEN = 
+
+// Leanplum SDK Variables
+LEANPLUM_ENVIRONMENT = development
```
Comment (st3fan) [inline Client/Configuration/Fennec.xcconfig]: I don't think we have a better way to do this. Ideally we do not put any API tokens in Git. For the production builds we must have them in secret BuddyBuild variables that can be replaced in a `buddybuild_prebuild.sh` script. Let me know if I can help with that.

File: Client/Frontend/Browser/TabManager.swift [hunk 21]
```diff
@@ -453,14 +453,39 @@ class TabManager: NSObject {
...
             })
         }
 
         delegates.forEach { $0.get()?.tabManagerDidRemoveAllTabs(self, toast: toast) }
     }
-    
+
+    func showFocusPromoToast() {
+        // add check for event stuff
+        guard let shouldShowFocusUI = LeanplumIntegration.sharedInstance.getBoolVariableFromServer(key: "shouldShowFocusUI"), let focus = NSURL(string: "firefox-focus://"), let klar = NSURL(string: "firefox-klar://"), !UIApplication.shared.canOpenURL(focus as URL) && !UIApplication.shared.canOpenURL(klar as URL) && shouldShowFocusUI else {
```
Comment (st3fan) [inline Client/Frontend/Browser/TabManager.swift]: There is so much logic here on one line. How about a `shouldShowFocusUI()` function?

File: Client/Frontend/Browser/TabManager.swift [hunk 22]
```diff
@@ -453,14 +453,39 @@ class TabManager: NSObject {
...
+            return
+        }
+
+        var toast: LeanplumPromoButtonToast
+        let templateDict = LeanplumIntegration.sharedInstance.getTemplateDictionary()!
```
Comment (st3fan) [inline Client/Frontend/Browser/TabManager.swift]: Please change this into an `if let` so that we can get rid of the explicit unwrap.

File: Client/Frontend/Browser/TabManager.swift [hunk 23]
```diff
@@ -453,14 +453,39 @@ class TabManager: NSObject {
...
+
+        LeanplumIntegration.sharedInstance.forceContentUpdate()
+        toast = LeanplumPromoButtonToast(labelText: templateDict["Template Text"]! , buttonText: templateDict["Button Text"]!, colorText: templateDict["Hex Color String"]!, completion: { buttonPressed in
+            guard let url = NSURL(string: templateDict["Deep Link"]!) else {
```
Comment (st3fan) [inline Client/Frontend/Browser/TabManager.swift]: Same here with the explicit unwraps. The assumption is probably that they always work but it is better to be safe and always use `if let`.

File: Client/Frontend/Browser/TabManager.swift (hunk 21 above)
Comment (farhanpatel) [inline Client/Frontend/Browser/TabManager.swift]: This could also move into the LeanplumIntegration instead.

File: Client/Frontend/Browser/TabTrayController.swift [hunk 24]
```diff
@@ -497,57 +497,57 @@ class TabTrayController: UIViewController {
 
     func didLongPressTab(_ gesture: UILongPressGestureRecognizer) {
         switch gesture.state {
-            case .began:
```
Comment (st3fan) [inline Client/Frontend/Browser/TabTrayController.swift]: What are these changes? Is this MMA related?

File: Client/Frontend/Browser/TabManager.swift (hunk 21 above)
Comment (Sdaswani) [inline Client/Frontend/Browser/TabManager.swift]: Not sure why it should move to the LeanplumIntegration @farhanpatel ? I think that should serve as a wrapper for the LP SDK only, so not sure why this logic (which concerns the use of LP) should move there? I don't think it will used by other users of LeanplumIntegration so no benefit putting it there. Overall though, not a huge deal. :)

File: Client/Frontend/Browser/TabTrayController.swift (hunk 24 above)
Comment (Sdaswani) [inline Client/Frontend/Browser/TabTrayController.swift]: No, I don't believe - more of the whitespace changes.

File: Client/Frontend/Browser/TabManager.swift (hunk 21 above)
Comment (farhanpatel) [inline Client/Frontend/Browser/TabManager.swift]: Yea, you are right its not a huge deal.  But I usually try to keep logic isolated. The file TabManager is related to managing tabs. So logic concerning LeanPlum feels like it belongs more in LeanPlumIntegration than in TabManger. 
Also think of the following case. There might be other places where you want to show a focus alert/toast. You’ll also need to check if focus is installed there as well. Instead of having to write this line again you can just do 
`LeanPlumIntegration.sharedInstance.shouldShowFocusInfo()`

Comment (Sdaswani) [inline Client/Frontend/Browser/TabManager.swift]: OK, separation of concerns noted :) . I would then suggest we put it in a third, new place, as whether focus is installed or not can be of general concern to a range of parties, including MMA. Maybe in a class in the Shared directory?

File: Client/Configuration/Fennec.xcconfig (hunk 20 above)
Comment (bkmunar) [inline Client/Configuration/Fennec.xcconfig]: Yes I need help with this, let's do this in a separate branch or something. Contact me about this!

**Comment by bkmunar** [comment]
ITS A MIRACLE!!!! thanks @farhanpatel

File: Client/Application/AppDelegate.swift [hunk 25]
```diff
@@ -158,11 +158,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UIViewControllerRestorati
         
         adjustIntegration = AdjustIntegration(profile: profile)
 
-        // We need to check if the app is a clean install to use for
-        // preventing the What's New URL from appearing.
-        if getProfile(application).prefs.intForKey(IntroViewControllerSeenProfileKey) == nil {
-            getProfile(application).prefs.setString(AppInfo.appVersion, forKey: LatestAppVersionProfileKey)
```
Comment (farhanpatel) [inline Client/Application/AppDelegate.swift]: I think this is not exactly related to Leanplum. Are you sure this should be removed?

**Comment by farhanpatel** [comment]
Also. Wait for Danger to finish. In case there are any lints I missed.

File: Client/Application/AppDelegate.swift (hunk 25 above)
Comment (bkmunar) [inline Client/Application/AppDelegate.swift]: woops forgot to transfer that over!

Comment (bkmunar) [inline Client/Application/AppDelegate.swift]: wait, i did transfer it over! it's in leanplum integration... should i separate it? might be redundant

Comment (farhanpatel) [inline Client/Application/AppDelegate.swift]: Is this related to leanplum though? wasnt this code here before leanplum?

Comment (bkmunar) [inline Client/Application/AppDelegate.swift]: it's not, but that's how the events are triggered... maybe i should just take it back out of LeanplumIntegration?

Comment (farhanpatel) [inline Client/Application/AppDelegate.swift]: Yea you should take it out. If we remove leanplum we want to make sure we dont accidently remove this as well

**Comment by bkmunar** [comment]
There, should be all done!

File: Client/Configuration/Fennec.xcconfig (hunk 20 above)
Comment (Sdaswani) [inline Client/Configuration/Fennec.xcconfig]: https://bugzilla.mozilla.org/show_bug.cgi?id=1359643

**Comment by farhanpatel** [comment]
Just the last 2 nits. and you should be good!
